lib/ConsumerImpl.cc

```diff
@@ -37,15 +37,14 @@ namespace pulsar {
 
 DECLARE_LOG_OBJECT()
 
-ConsumerImpl::ConsumerImpl(const ClientImplPtr client, const std::string& topic,
-                           const std::string& subscriptionName, const ConsumerConfiguration& conf,
-                           bool isPersistent,
+ConsumerImpl::ConsumerImpl(ClientImpl& client, const std::string& topic, const std::string& subscriptionName,
```

Comment:
Why do we need to change from storing a `weak_ptr` to a reference?

The problem with this is that if you keep using a producer/consumer after the client is already destroyed, you would have a segfault. With the `weak_ptr` we are force to always validating it before using it.

Reply:
> The problem with this is that if you keep using a producer/consumer after the client is already destroyed, you would have a segfault.

Actually not. See `TEST_P(ClientCloseTest, testShutdown)`.

> It seems this PR is included more changes than just the part of using a map to keep track of producers/consumers disposal.

And yes. I think it needs another PR for this refactor. For this PR, I will revert these unnecessary changes soon.

Reply:
> Actually not. See TEST_P(ClientCloseTest, testShutdown).

I think it's probably that most method have already state checks at the beginning. 

(Also I'm seeing many places where are currently doing `client_->lock()` and not checking if we got a valid shared_ptr or an empty one...)

> And yes. I think it needs another PR for this refactor. For this PR, I will revert these unnecessary changes soon.

👍 👍

Reply:
> I think it's probably that most method have already state checks at the beginning.

Yes. If the associated `Client` is expired, the producer or consumer should not be able to do anything except a `ResultAlreadyClosed`.